tests/models/test_taskinstance.py

```diff
@@ -17,15 +17,16 @@
 # under the License.
 
 import datetime
+import math
 import os
 import signal
-import time
 import urllib
 from tempfile import NamedTemporaryFile
 from typing import List, Optional, Union, cast
 from unittest import mock
 from unittest.mock import call, mock_open, patch
 
+import numpy as np
```

Comment:
This library is not an Airflow dependency.  If we want to use it for core testing, it should be installed with `deve` extral.
https://github.com/apache/airflow/blob/main/setup.py#L495
Or maybe you have an idea how to write this test without using this library?

Reply:
Yeah, it is not strictly necessary. I was just trying to produce the closest boundary case 🙂 

In general, is there a preference of whether to add (not strictly necessary) libraries to the `devel` extra?
https://github.com/apache/airflow/blob/3f65a48bcedfab6a50e24ecac710170aab6f51bc/tests/models/test_taskinstance.py#L526-L532

Reply:
Actually, I just saw #12500. It doesn't seem necessary to require `numpy` just for this test; I will change it. But do you know why the test still passed in the CI environment?

Reply:
We have a lot of transitive dependencies. In this case, we have installed `numpy` on CI, because. it is required by `apache-beam`, `pandas` and `pyarrow`.
```
$ pipdeptree -r -p numpy
numpy==1.19.5
  - apache-beam==2.32.0 [requires: numpy>=1.14.3,<1.21.0]
  - pandas==1.1.5 [requires: numpy>=1.15.4]
    - pandas-gbq==0.14.1 [requires: pandas>=0.20.1]
    - scrapbook==0.5.0 [requires: pandas]
  - pyarrow==4.0.1 [requires: numpy>=1.16.6]
    - apache-beam==2.32.0 [requires: pyarrow>=0.15.1,<5.0.0]
    - scrapbook==0.5.0 [requires: pyarrow]
```
Airflow does not depend on these dependencies, so if we want to use this library, we should add it.
```
$ curl -s https://raw.githubusercontent.com/apache/airflow/main/setup.{py,cfg} | grep -C 10 numpy | wc -l
0
```
> In general, is there a preference of whether to add (not strictly necessary) libraries to the devel extra?

In this case, we should add this library to `devel`, because itis needed for core.

extra | command | description
-- | -- | --
devel | pip install 'apache-airflow[devel]' | Minimum dev tools requirements (without providers)
devel_hadoop | pip install 'apache-airflow[devel_hadoop]' | Same as devel + dependencies for developing the Hadoop stack
devel_all | pip install 'apache-airflow[devel_all]' | Everything needed for development (devel_hadoop + providers)
devel_ci | pip install 'apache-airflow[devel_ci]' | All dependencies required for CI build.

See: http://airflow.apache.org/docs/apache-airflow/stable/extra-packages-ref.html#bundle-extras

Reply:
I also support the idea of getting rid of this dependency, but if you don't find another solution then it's okay.

Reply:
Thanks for the explanation! Yes, I found a solution where the difference is practically negligible, so I will just change it and remove `numpy` from the test.